feat: Add Socket Security integrations and fix scan failure detection

dacoburn opened the pull request:
Socket Reachability & SCA Integration:
- Add Socket reachability scanning with configurable org parameter
- Add Socket SCA scanning with configurable files parameter
- Create connector classes for Socket and SocketSCA tools
- Add environment variable support for Socket API keys

Bug Fix - Scan Failure Detection:
- Fix critical issue where Socket SCA scan failures weren't causing build failures
- Implement proper JSON extraction from mixed log/JSON output using regex pattern
- Add file path resolution using TEMP_OUTPUT_DIR environment variable
- Add explicit scan failure detection logic that exits with code 1
- Create SocketSCAEvent class for proper event handling in console output

Technical Improvements:
- Update entrypoint.sh with Socket CLI integration and JSON parsing
- Enhance socket_external_tools_runner.py with scan failure detection
- Add comprehensive error handling and logging
- Update action.yml with new Socket-related input parameters
- Add Python version management with .python-version and pyproject.toml
- Include test files for validating Socket SCA failure scenarios

The fix ensures that when Socket SCA returns scan_failed: true with critical alerts, the system properly exits with error code 1 and displays appropriate error messaging instead of incorrectly reporting 'No issues detected'.

dacoburn committed:
feat: migrate to uv package manager and implement version management system

- Replace pip with uv for faster dependency resolution and modern Python packaging
- Add pyproject.toml with project configuration and dependencies
- Update Dockerfile to use uv sync --frozen for reproducible builds
- Implement comprehensive version management system:
  - Create src/version.py for runtime version imports
  - Add .hooks/version-check.py pre-commit hook with auto-bump functionality
  - Support both patch and dev version bumping (.devN)
  - Sync versions between pyproject.toml and version.py automatically
  - Include PyPI conflict detection
- Update documentation with uv installation and usage instructions
- Add version management setup and usage to README
- Remove requirements.txt (replaced by pyproject.toml + uv.lock)
- Update version to 1.0.18
- Maintain full compatibility with existing Docker workflow and GitHub Actions

Dependencies: All previous dependencies preserved in pyproject.toml

dacoburn committed:
Implemented pre commit hook

dacoburn committed:
Add Socket NPM & Python CLI to Docker container

- Install Socket NPM CLI globally via npm
- Install Socket Python CLI via uv tool install
- Both CLIs will be available for security scanning workflows
- Version bumped to 1.0.20

dacoburn committed:
Added version file

dacoburn committed:
feat: Add Socket Security integrations and fix scan failure detection

Socket Reachability & SCA Integration:
- Add Socket reachability scanning with configurable org parameter
- Add Socket SCA scanning with configurable files parameter
- Create connector classes for Socket and SocketSCA tools
- Add environment variable support for Socket API keys

Bug Fix - Scan Failure Detection:
- Fix critical issue where Socket SCA scan failures weren't causing build failures
- Implement proper JSON extraction from mixed log/JSON output using regex pattern
- Add file path resolution using TEMP_OUTPUT_DIR environment variable
- Add explicit scan failure detection logic that exits with code 1
- Create SocketSCAEvent class for proper event handling in console output

Technical Improvements:
- Update entrypoint.sh with Socket CLI integration and JSON parsing
- Enhance socket_external_tools_runner.py with scan failure detection
- Add comprehensive error handling and logging
- Update action.yml with new Socket-related input parameters
- Add Python version management with .python-version and pyproject.toml
- Include test files for validating Socket SCA failure scenarios

The fix ensures that when Socket SCA returns scan_failed: true with critical
alerts, the system properly exits with error code 1 and displays appropriate
error messaging instead of incorrectly reporting 'No issues detected'.

dacoburn committed:
Removed test files

dacoburn committed:
WIP: Current changes before merging jira support branch

dacoburn committed:
Merge doug/add-jira-support into doug/add-socket-sca-and-jira

- Resolved conflicts in Dockerfile (kept uv tool install approach)
- Resolved version conflicts and auto-bumped to 1.0.23
- Successfully merged JIRA support features

mtorp commented on debug_test.py:
```diff
@@ -0,0 +1,79 @@
+import tempfile
+
+# Add the src directory to the Python path
+sys.path.insert(0, '/Users/douglascoburn/socket.dev/github/security-tools/src')
```
Do you want to commit the absolute paths in this file?

dacoburn replied:
Didn't mean to include that file

mtorp reviewed (approved):
✅

dacoburn committed:
Merged changes from other branch

dacoburn committed:
feat: Major security wrapper improvements - S3 integration, PURL standardization, and consolidated facts processing

🔧 Core Infrastructure:
- Add light-s3-client 0.0.30 dependency for S3-compatible storage
- Implement S3 upload/download with change detection for .socket.facts.json
- Update Docker container structure with proper working directories
- Fix entrypoint.sh duplicate declarations and consolidation logic

📊 Data Format Standardization:
- Standardize tool types to format like 'sast-bandit', 'secrets-trufflehog', 'container-trivy'
- Implement proper PURL format: pkg:ecosystem/name@version?type=tool-type
- Add PURL creation for Socket components that initially lack purl fields
- Update version to 1.0.25 across all components

🔄 Socket Facts Consolidation:
- Add _process_socket_vulnerabilities method to create alerts from Socket reachability data
- Implement comprehensive security tool result consolidation into .socket.facts.json
- Add new_alerts field with S3-based change detection to prevent duplicate processing
- Support for all security tools: Bandit, Gosec, ESLint, Trufflehog, Trivy, Socket SCA, Socket Reachability

🎯 Plugin Integration Improvements:
- Update Jira plugin with enhanced table format including Tool and Source columns
- Add _extract_tool_from_purl() and _clean_purl_source() methods to all plugins
- Implement send_consolidated_security_alerts() across Jira, Slack, Teams, Webhook plugins
- Fix Jira authentication and project access with better error handling

🐛 Bug Fixes:
- Fix Socket SCA JSON extraction from timestamped output
- Resolve merge conflict in socket_sca connector
- Fix Trufflehog file path normalization and line number extraction
- Improve error handling and debug output across all components

This update establishes a robust foundation for security tool integration with proper data standardization, S3-based state management, and comprehensive plugin support for external integrations.

dacoburn committed:
refactor: implement Socket Facts schema constraints for component types

- Socket Reachability/SCA components now use ecosystem types (npm, pypi, etc.)
- SAST, secrets, and container scanning use generic component type
- Group alerts by file using qualifiers for shared context
- Move code blocks into alert properties instead of separate fields
- Multiple alerts per file consolidated into single components
- Preserve Socket vulnerability data with ecosystem-specific typing

This aligns the consolidated facts format with the injectable alert
artifact schema constraints while maintaining backwards compatibility
for Socket tooling that expects ecosystem-typed components.

dacoburn committed:
feat: Add Socket SCA integration with multi-platform alert notifications

- Add Socket SCA (Software Composition Analysis) scanning capability
- Implement consolidated notification system with Jira, Slack, Teams, and Webhook plugins
- Add comprehensive alert processing pipeline for external security tools
- Update action.yml with new input parameters for SCA and notification configs
- Enhance entrypoint.sh with Socket SCA environment variable handling
- Create modular plugin architecture for extensible notification backends
- Add Socket SCA connector with proper alert filtering and severity handling
- Update README with integration configuration examples and setup instructions
- Improve consolidated facts processing for multi-tool security alerts

This adds a complete notification ecosystem for security findings, allowing teams to receive alerts through their preferred channels while maintaining the existing Socket facts format.
